addons/web/controllers/main.py

```diff
@@ -1538,6 +1540,132 @@ def from_data(self, fields, rows):
         fp.close()
         return data
 
+
+class XMLExport(ExportFormat, http.Controller):
+    """Export records to XML
+
+    Export records in Odoo XML data files format. The exported fields of a model
+    are the union of all selected fields of that model.
+    """
+
+    @http.route('/web/export/xml', type='http', auth="user")
+    @serialize_exception
+    def index(self, data, token):
+        return self.base(data, token)
+
+    @property
+    def content_type(self):
+        return 'text/xml; charset=utf-8'
+
+    def filename(self, base):
+        return base + '.xml'
+
+    def extract_types(self, field_names, model):
+        """Extract models data for given fields
+
+        Update object types and comodels corresponding to a list of fields.
+
+        Args:
+            field_names: A list of fields.
+        """
+
+        models_types = defaultdict(dict)
+        models_comodels = defaultdict(dict)
+
+        for tokens in (field_name.split('/') for field_name in field_names):
+            cur_model = model
+            for tok in tokens:
+                if tok in ("id", ".id"):
+                    continue
+                if tok not in models_types[cur_model]:
+                    attrs = request.env[cur_model].fields_get([tok], attributes=['type', 'relation'])[tok]
+                    if attrs['type'] in ("one2many", "many2many", "many2one"):
+                        models_types[cur_model][tok] = attrs['type']
+                        models_comodels[cur_model][tok] = attrs['relation']
+                    else:
+                        models_types[cur_model][tok] = attrs['type']
+                if tok in models_comodels[cur_model]:
+                    cur_model = models_comodels[cur_model][tok]
+
+        self.models_types = models_types
+        self.models_comodels = models_comodels
+
+    def export_record(self, record, model):
+        """Add a record to the root element
+
+        Add a record as an xml element to the xml root element. Can call
+        itself to add related records. If a record as already been
+        exported, the function only return the xml_id.
+
+        Args:
+            record: A recordset containing one record.
+            model: The model of the record to add.
+
+        Returns:
+            The xml_id of the added or existing record.
+        """
+
+
+        if record.id in self.seen_ids[model]:
+            return self.seen_ids[model][record.id]
+
+        xml_id = record._BaseModel__export_xml_id()
+        start_no_xml_id = '__export__.'
+        if xml_id.startswith(start_no_xml_id):
+            name = record.name_get()
+            if name:
+                xml_id = xml_id[len(start_no_xml_id):]+'_'+re.sub('[^a-zA-Z0-9-]', '_', name[0][1])
+            else:
+                xml_id = xml_id[len(start_no_xml_id):]
```

Comment:
Might be time to create "official" internal APIs for querying and generating nice external ids /cc @odony